075-connector-offsets-support.md

```diff
@@ -0,0 +1,270 @@
+[KIP-875][kip] adds new endpoints to let the user do this more easily.
+There are multiple reasons a user might want to manage offsets, for example (as listed in the KIP):
+* Resetting the offsets for a connector while iterating rapidly in a development environment (so that the connector does not have to be renamed each time)
+* Viewing the in-memory offsets for source connectors on the cluster in order to recover from accidental deletion of the Connect source offsets topic (this is currently possible with the config topic, and can be a life-saver)
+* Monitoring the progress of running connectors on a per-source-partition basis 
+* Skipping records that cause issues with a connector and cannot be addressed using existing error-handling features
+
+## Proposal
+
+Strimzi should include mechanisms to let users manage offsets when utilising the KafkaConnector resources.
+This proposal suggests introducing a new annotation on the KafkaConnector resource to trigger each operation.
+The operations are:
+* List offsets.
+* Alter offsets.
+* Reset offsets.
+
+The list offsets operation uses a ConfigMap for the output.
+The alter offsets operation uses a ConfigMap for the input.
+
+### Request/Response from the API endpoints
+
+The response body received when using the GET endpoint, and the request body used when calling the PATCH endpoint have identical structures to allow users to easily fetch and then alter offsets.
+The structure of the body is different for source and sink connectors.
+This is because sink connectors use regular Kafka offsets, whereas source connectors determine their own structure for offsets.
+This difference is reflected in the approaches outlined in this proposal.
+
+Source connector format is:
+```
+{
+  "offsets": [
+    {
+      "partition": {
+        // Connector-defined source partition
+      },
+      "offset": {
+        // Connector-defined source offset
+      }
+    }
+  ]
+}
+```
+
+Sink connector format is:
+```
+{
+  "offsets": [
+    {
+      "partition": {
+        "kafka_topic": // Kafka topic
+        "kafka_partition": // Kafka partition
+      },
+      "offset": {
+        "kafka_offset": // Kafka offset
+      }
+    }
+  ]
+}
+```
+
+### New annotation
+
+All three actions will make use of a new annotation called `strimzi.io/connectorOffsets`.
+The possible values will be `list`, `alter` and `reset`.
+
+### Listing offsets
+
+To list the current offsets the user will annotate the KafkaConnector resource with the new annotation `strimzi.io/connectorOffsets` set to `list`.
+After the annotation is added, on the next reconciliation the operator will fetch the current offsets and create a Kubernetes ConfigMap containing the JSON response.
+The operator will add an owner reference to the Kubernetes ConfigMap pointing to the KafkaConnector resource that was annotated.
+If the ConfigMap already exists the operator will replace the data with the updated data and not add an owner reference.
```

Comment:
Assuming by "the data" you mean "the `data`", then we're saying we'll obliterate any other keys in the ConfigMap. It might be worth calling this out, even if only by monospacing the `data`.

Reply:
Yes, I've adjusted the wording